Sources/GRPC/ConnectionPool/PoolManager.swift

```diff
@@ -0,0 +1,314 @@
+  /// transport.
+  internal struct PooledStreamChannel {
+    /// The future `Channel`.
+    var futureResult: EventLoopFuture<Channel>
+
+    /// The `EventLoop` that the `Channel` is using.
+    var eventLoop: EventLoop {
+      return self.futureResult.eventLoop
+    }
+  }
+
+  /// Make a stream and initialize it.
+  ///
+  /// - Parameters:
+  ///   - preferredEventLoop: The `EventLoop` that the stream should be created on, if possible. If
+  ///       a pool exists running this `EventLoop` then it will be chosen over all other pools,
+  ///       irregardless of the load on the pool. If no pool exists on the preferred `EventLoop` or
+  ///       no preference is given then the pool with the most streams available will be selected.
+  ///       The `EventLoop` of the selected pool will be the same as the `EventLoop` of
+  ///       the `EventLoopFuture<Channel>` returned from this call.
+  ///   - deadline: The point in time by which the stream must have been selected. If this deadline
+  ///       is passed then the returned `EventLoopFuture` will be failed.
+  ///   - logger: A logger.
+  ///   - initializer: A closure to initialize the `Channel` with.
+  /// - Returns: A `PoolStreamChannel` indicating the future channel and `EventLoop` as that the
+  ///     `Channel` is using. The future will be failed if the pool manager has been shutdown,
+  ///     the deadline has passed before a stream was created or if the selected connection pool
+  ///     is unable to create a stream (if there is too much demand on that pool, for example).
+  internal func makeStream(
+    preferredEventLoop: EventLoop?,
+    deadline: NIODeadline,
+    logger: GRPCLogger,
+    streamInitializer initializer: @escaping (Channel) -> EventLoopFuture<Void>
+  ) -> PooledStreamChannel {
+    let reservation = self.lock.withLock {
+      self.state.reserveStream(
+        preferringPoolWithEventLoopID: preferredEventLoop.map { EventLoopID($0) }
+      )
+    }
+
+    switch reservation {
+    case let .success(poolIndex):
+      let channel = self.pools[poolIndex.value].makeStream(
+        deadline: deadline,
+        logger: logger,
+        initializer: initializer
+      )
+      return PooledStreamChannel(futureResult: channel)
+
+    case let .failure(error):
+      let eventLoop = preferredEventLoop ?? self.group.next()
+      return PooledStreamChannel(futureResult: eventLoop.makeFailedFuture(error))
+    }
+  }
+
+  // MARK: Shutdown
+
+  /// Shutdown the pool manager and all connection pools it manages.
+  internal func shutdown(promise: EventLoopPromise<Void>) {
+    let (action, pools): (PoolManagerStateMachine.ShutdownAction, [ConnectionPool]?) = self.lock
+      .withLock {
+        let action = self.state.shutdown(promise: promise)
+
+        switch action {
+        case .shutdownPools:
+          // Clear out the pools; we need to shut them down.
+          let pools = self.pools
+          self.pools.removeAll()
```

Comment:
Suggest using `keepingCapacity: true` to avoid Array thinking we want to free memory.